Allow for redistribution motions in CPhysicalNLJoins in the general case

danhoffman699 opened the pull request:
Previously, `CPhysicalNLJoin` required a replicated motion in the general case.
This forced a broadcast motion in more cases than required. This commit allows
for the inner child to match the distribution of the outer child via a
redistriution motion.

```
create table foo (a int, b int);
create table bar (c int, d int);

insert into foo values (1, 1), (2, 2), (3, 3);
insert into bar values (3, 3), (2, 2), (1, 1);

analyze foo;
analyze bar;

set optimizer_enable_hashjoin = 0;
explain select * from foo left join bar on (a = d);
```

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

danhoffman699 commented:
NOTE: As written, `CPhysicalInnerNLJoin::Ped` is equivalent to `CPhysicalNLJoin::Ped`. If there isn't feedback specific to the `CPhysicalNLJoin::Ped` implementation, then I may remove the `CPhysicalInnerNLJoin::Pad` implementation as it would be redundant.

danhoffman699 commented:
Removed unused distribution optimization case and replaced it with an Any distribution on the outer side and a Replicated distribution on the inner (the original optimization case opportunistically created a Hashed distribution on the inner if the current distribution of the outer is hashed, among other constraints).

danhoffman699 commented:
I ran a benchmark suite and explicitly adding any-replicated alongside any-hash/replicated (choose hash if applicable, otherwise choose replicated) adds ~100ms to the query check time with no plan changes.

chrishajas commented:
> I ran a benchmark suite and explicitly adding any-replicated alongside any-hash/replicated (choose hash if applicable, otherwise choose replicated) adds ~100ms to the query check time with no plan changes.

What is the total optimization time? ~100ms does seem like a large increase unfortunately.

danhoffman699 commented:
The baseline time is ~1500ms, so its an increase of ~7%. No plan changes were observed between the baseline/master, not generating replicated when hashed can be applied, and generating hashed and replicated (at least not in the benchmarks ran).

danhoffman699 commented:
I've removed the explicit replicated distribution spec generation if hashed is sufficient (no plan changes were observed when we had the two and it introduced a meaningful increase to optimization time). I've rebased the branch to resolve conflicts with master. I also ran clang-fmt so this (should) properly pass the CI

chrishajas reviewed (commented):
This will also change LeftSemiNL/LeftAntiSemiNL joins' Ped(). Is that ok? In either case, we should add tests if we don't have them already.

I agree that we should minimize the duplicate code here if possible too.

bhuvnesh2703 commented:
Replaced by https://github.com/greenplum-db/gpdb/pull/12586
